indexer/tasks/src/multiera/multiera_projected_nft.rs

```diff
@@ -110,107 +115,99 @@ async fn handle_projected_nft(
+
+            let output_address =
+                cardano_multiplatform_lib::address::Address::from_bytes(output_address)
+                    .map_err(|err| DbErr::Custom(format!("cml can't parse address: {}", err)))?;
+            let output_payment_cred = match output_address.payment_cred() {
                 None => continue,
                 Some(pk) => pk,
             };
 
-            if output_payment_cred != config_payment_cred {
+            if output_payment_cred != projected_nft_contract_payment_cred {
+                // the output doesn't relate to projected nft contract -> we don't care about it
                 continue;
             }
 
-            let output_model = match outputs_map.get(&(output_index as i32)) {
-                None => {
-                    return Err(DbErr::RecordNotFound(format!(
-                        "can't find output with index {output_index} of tx {}",
-                        cardano_transaction.id
-                    )))
-                }
-                Some(output) => output.clone(),
-            };
+            // current output details
+            let output_model = outputs_map
+                .get(&(output_index as i32))
+                .ok_or(DbErr::RecordNotFound(format!(
+                    "can't find output with index {output_index} of tx {}",
+                    cardano_transaction.id
+                )))?
+                .clone();
+
+            // parse the state and fetch projected nft details
+            let projected_nft_data =
+                extract_operation_and_datum(output, output_model, &partial_withdrawals_inputs);
+
+            // if projected nft data is unlocking output that is created via partial withdrawal
+            // then we reduce associated partial withdrawal input balance by the amounts that are being unlocked
+            // this way we will be able to find the corresponding lock output that will have the rest of the balance
+            handle_partial_withdraw(&projected_nft_data, &mut partial_withdrawals_inputs)?;
+
+            projected_nft_outputs.push(projected_nft_data);
+        }
 
-            let projected_nft_data = extract_operation_and_datum(output);
+        find_lock_outputs_for_corresponding_partial_withdrawals(
+            &mut projected_nft_outputs,
+            &mut partial_withdrawals_inputs,
+        )?;
 
-            for asset in output.non_ada_assets() {
+        if !partial_withdrawals_inputs.is_empty() {
+            return Err(DbErr::Custom(format!("Partial withdrawals must be empty at the end of projected nft processing, while contains: {}", partial_withdrawals_inputs.keys().map(hex::encode).fold(String::new(), |acc, key| format!("{acc},{key}")))));
+        }
+
+        for output_data in projected_nft_outputs.into_iter() {
+            for (asset_name, asset_value) in output_data.non_ada_assets.into_iter() {
                 queued_projected_nft_records.push(entity::projected_nft::ActiveModel {
```

Comment:
It feels weird to me that the abstraction is per-asset. I feel like the correct abstraction is just tracking which UTXO are projected and that's it. That is to say, is there a downside to having a single row per UTXO (instead of one row per asset) and just storing the CBOR of the assets in the row so it can be decoded later?

Had we done it this way, we could a smaller database and would not longer need the rows `plutus_datum`, `amount`, `asset`, `tx_id`

Maybe there is a reason we chose to do it this way instead? This way seems more error prone, has less information and takes more space
I think it's not that big of a deal since probably the size increase in the database caused by this is not big and it doesn't affect the end-user REST api interface

Reply:
From the db perspective single row per utxo could work as well, it's more the matter of when and where to do the parsing and how convenient is the search. Main pros of row-per-utxo approach are: 
* all assets are in one place and you never loose them
* all data is stored once (e.g. plutus_data)

It has cons as well:
* no search per asset
* cbor is not human readable and the clients will need to parse cbor
Besides that, in any case on the paima side we will care about which assets have which status (and who owns them), not which utxos. Otherwise you'll need to fetch all of them for an address and parse their cbor as well

So I chose the row per asset approach just because it could potentially cover more cases and on the client side you can not care about what cbor is and how to parse it.  Still big advantage of single row per utxo approach is the fact that you care about status of the whole utxo and not the utxo - asset pairs. 

For current solution I'd say I see one more improvement: having `policy id` and `asset name` instead of `asset`, so it's easier to query by particular collection. If you insist we have single row per utxo we can still do it, but it will reduce the readability and search capabilities